fish_prompt.fish

```diff
@@ -81,9 +81,8 @@ function fish_prompt
     end
 
     # Check if there is an upstream configured
-    set -l has_upstream (command git rev-parse --abbrev-ref @'{u}' ^/dev/null)
-
-    if test -n "$has_upstream"
+    command git rev-parse --abbrev-ref '@{upstream}' >/dev/null ^&1
```

Comment:
1. `set variable (command)` sets `variable` with the output of `command`. It is useful in tons of cases, but what really want in this case is to get the return code from the `git` sub-command. Hence it is not useful to run in a sub-command.
We run the command, then get the return code with the special variable `$status`.

2. From [`git rev-parse`]:
> **<branchname>@{upstream}, e.g. master@{upstream}, @{u}**
> The suffix @{upstream} to a branchname (short form <branchname>@{u}) refers to the branch that the branch specified by branchname is set to build on top of (configured with branch.<name>.remote and branch.<name>.merge). A missing branchname defaults to the current one.

  So `@{upstream}` or `@{u}` is a symbolic link to the base repository (what we usually call "origin" in git) of the current git branch. I believe that `@{u}` was a later addition to `git`, and also it's more explicit, and that's way I made it a habit to use `@{upstream}`.


3. `>` redirects the `stdout` pipe (`stdout` is an alias to file descriptor 1).
`/dev/null` is a special *nix "device" that simply ignores everything you write to it.
`^` redirects the `stderr` pipe (alias to file descriptor 2).
`&1` means to file descriptor (fd) 1.
When you put it all together, `>/dev/null ^&1` means "redirect all error output to regular output, and oh look! all regular output is ignored". Bottom line: this command will not write anything to the screen.

[`git rev-parse`]: https://git-scm.com/docs/git-rev-parse